server/controllers/finance/reports/debtors/openDebtors.js

```diff
@@ -35,112 +62,114 @@ function build(req, res, next) {
-          UNION (
-            SELECT entity_uuid, debit_equiv, credit_equiv FROM general_ledger
-          )
-        ) AS c
-        WHERE c.entity_uuid IN (SELECT patient.debtor_uuid FROM patient)
-        GROUP BY c.entity_uuid
-        HAVING balance > 0
-      ) AS ledger ON ledger.entity_uuid = patient.debtor_uuid
-    GROUP BY patient.debtor_uuid
-    ORDER BY ${ordering};
+// ONLY select rows with an entity
+// ONLY show debtors with a debt above 0
+function buildDebtQuery(showDetailedView, source, dateCondition, ordering) {
+  // Include complex parameters depending on detailed view Boolean requirements
+  const complexParameters = showDetailedView ?
+    ', MAX(invoice.date) as latestInvoiceDate, MAX(cash.date) as latestCashDate ' :
+    '';
+  const complexJoin = showDetailedView ?
+    'LEFT JOIN invoice on reference_uuid = invoice.uuid LEFT JOIN cash on reference_uuid = cash.uuid ' :
+    '';
+
+  // Include all balance and debtor information by default
+  const query = `
+    SELECT patient.display_name, entity_map.text as reference,
+      SUM(debit_equiv - credit_equiv) as balance ${complexParameters}
+    FROM ${source}
+    JOIN patient on entity_uuid = patient.debtor_uuid
+    LEFT JOIN entity_map on entity_map.uuid = entity_uuid
+    ${complexJoin}
+    WHERE entity_uuid IS NOT NULL
+    ${dateCondition}
+    GROUP BY entity_uuid
+    HAVING SUM(debit_equiv - credit_equiv) > 0
+    ORDER BY ${ordering}
   `;
 
-  /*
-   * Aggregate SQL for totalling all debts up to the present
-   */
-  const aggregateSql = `
-    SELECT COUNT(a.entity_uuid) AS numDebtors, SUM(a.debit) AS debit, SUM(a.credit) AS credit, SUM(a.balance) AS balance
-    FROM(
-      SELECT c.entity_uuid, SUM(c.debit_equiv) AS debit, SUM(c.credit_equiv) AS credit,
-        SUM(c.debit_equiv - c.credit_equiv) AS balance
-      FROM (
-        (
-          SELECT entity_uuid, debit_equiv, credit_equiv FROM posting_journal)
-        UNION (
-          SELECT entity_uuid, debit_equiv, credit_equiv FROM general_ledger
-        )
-      ) AS c
-      WHERE c.entity_uuid IN (SELECT patient.debtor_uuid FROM patient)
-      GROUP BY c.entity_uuid
-      HAVING balance > 0
-    ) AS a;
-  `;
+  return query;
+}
 
-  const data = {};
+const orderMap = {
+  'transaction-date-asc' : 'trans_date ASC',
+  'transaction-date-desc' : 'trans_date DESC',
+  'debt-desc' : 'balance DESC, trans_date DESC',
+  'debt-asc' : 'balance ASC, trans_date DESC',
+  'patient-name-desc' : 'patient.display_name DESC',
+  'patient-name-asc' : 'patient.display_name ASC',
+};
 
-  // execute the query and build the report
-  return db.exec(debtorsSql)
-    .then((debtors) => {
-      data.debtors = debtors;
-      return db.one(aggregateSql);
-    })
-    .then((aggregates) => {
-      data.aggregates = aggregates;
-      return report.render(data);
-    })
-    .then(result => res.set(result.headers).send(result.report))
-    .catch(next)
-    .done();
+const DEFAULT_ORDER = 'debt-desc';
```

Comment:
A default order is set on the server but if nothing is set on the client it will use this as default. Potentially it would be a good user experience to set this default ordering on the client as well so that the user has an idea of which ordering will be used even if they don't select an order.